daemon/mgr/container_stats.go

```diff
@@ -0,0 +1,158 @@
+package mgr
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/alibaba/pouch/apis/types"
+
+	"github.com/containerd/cgroups"
+	containerdtypes "github.com/containerd/containerd/api/types"
+	"github.com/containerd/typeurl"
+	"github.com/docker/docker/pkg/ioutils"
+	"github.com/go-openapi/strfmt"
+	"github.com/sirupsen/logrus"
+)
+
+// StreamStats gets the stats from containerd side and send back to caller as a stream.
+func (mgr *ContainerManager) StreamStats(ctx context.Context, name string, config *ContainerStatsConfig) error {
+	c, err := mgr.container(name)
+	if err != nil {
+		return err
+	}
+
+	outStream := config.OutStream
+	if (!c.IsRunning() || c.IsRestarting()) && !config.Stream {
+		return json.NewEncoder(outStream).Encode(&types.ContainerStats{
+			Name: c.Name,
+			ID:   c.ID,
+		})
+	}
+
+	if c.IsRunning() && !config.Stream {
+		metrics, stats, err := mgr.Stats(ctx, name)
+		if err != nil {
+			return err
+		}
+		containerStat := toContainerStats(metrics.Timestamp, stats)
+		return json.NewEncoder(outStream).Encode(containerStat)
+	}
+
+	if config.Stream {
+		wf := ioutils.NewWriteFlusher(outStream)
+		defer wf.Close()
+		wf.Flush()
+		outStream = wf
+	}
+
+	enc := json.NewEncoder(outStream)
+
+	for {
+		select {
+		case <-ctx.Done():
+			logrus.Infof("context is cancelled when streaming stats of container %s", c.ID)
+			return nil
+		default:
+			logrus.Debugf("Start to stream stats of container %s", c.ID)
```

Comment:
since we don't have event listener for specific container right now, I think we should check the status of container before get stat. If the status has been changed from running to stopped, we should exit the loop.

Reply:
I checked the existing moby api and it doesn't exit and will wait for next restart.